HW2/P5/physics.pyx

```diff
@@ -0,0 +1,192 @@
+    return (deltax * vrelx + deltay * vrely) > 0.0
+
+
+cdef inline void collide(FLOAT *x1, FLOAT *v1,
+                         FLOAT *x2, FLOAT *v2) nogil:
+    cdef:
+        float x1_minus_x2[2]
+        float v1_minus_v2[2]
+        float change_v1[2]
+        float len_x1_m_x2, dot_v_x
+        int dim
+    # https://en.wikipedia.org/wiki/Elastic_collision#Two-dimensional_collision_with_two_moving_objects
+    for dim in range(2):
+        x1_minus_x2[dim] = x1[dim] - x2[dim]
+        v1_minus_v2[dim] = v1[dim] - v2[dim]
+    len_x1_m_x2 = x1_minus_x2[0] * x1_minus_x2[0] + x1_minus_x2[1] * x1_minus_x2[1]
+    dot_v_x = v1_minus_v2[0] * x1_minus_x2[0] + v1_minus_v2[1] * x1_minus_x2[1]
+    for dim in range(2):
+        change_v1[dim] = (dot_v_x / len_x1_m_x2) * x1_minus_x2[dim]
+    for dim in range(2):
+        v1[dim] -= change_v1[dim]
+        v2[dim] += change_v1[dim]  # conservation of momentum
+
+cdef void sub_update(FLOAT[:, ::1] XY,
+                     FLOAT[:, ::1] V,
+                     float R,
+                     int i, int count,
+                     UINT[:, ::1] Grid,
+                     float grid_size,
+                     float grid_spacing,
+                     omp_lock_t *locks) nogil:
+    cdef:
+        FLOAT *XY1, *XY2, *V1, *V2
+        int j, dim, x, y, grid_x, grid_y
+        float eps = 1e-5
+
+    # SUBPROBLEM 4: Add locking
+    XY1 = &(XY[i, 0])
+    V1 = &(V[i, 0])
+    #############################################################
+    # IMPORTANT: do not collide two balls twice.
+    ############################################################
+    # SUBPROBLEM 2: use the grid values to reduce the number of other
+    # objects to check for collisions.
+
+    grid_x = <int>(XY1[0]/grid_spacing)
+    grid_y = <int>(XY1[1]/grid_spacing)
+    
+    #We loop on a grid of 5x5 centered on the i object
+    for x in range(grid_x-2, grid_x+3):
+        
+        #We check that we are not our of boundaries in the x axis
+        if x >=0 and x < grid_size:           
+            for y in range(grid_y-2, grid_y+3): 
+                
+                #We check that we are not our of boundaries in the y axis
+                if y >= 0 and y < grid_size:
+                    
+                    #We get the value of the grid neighbor
+                    if (grid_x <x) or (grid_x == x and grid_y <y):
+                        j = Grid[x,y]
+
+                        #We check that the grid is not empty in that position and that we not checking the i object
+                        if j != -1 and j != i:
+                            #We get the coordinates and velocities of the neighbor
+                            XY2 = &(XY[j, 0])
+                            V2 = &(V[j, 0])
+
+                            if overlapping(XY1, XY2, R):
+                                
+                                #We lock the balls i and j, starting with the small index first to avoid deadlocks.
```

Comment:
Good way to check for deadlocking, I did something slightly different but once again I like your method better.